kv/kvserver: TestPromoteNonVoterInAddVoter failed

kv/kvserver.TestPromoteNonVoterInAddVoter [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21345318?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21345318?buildTab=artifacts#/) on master @ [06ffbe9546b8118e436f94eaff454aeb304046b2](https://github.com/cockroachdb/cockroach/commits/06ffbe9546b8118e436f94eaff454aeb304046b2):


```
=== RUN   TestPromoteNonVoterInAddVoter
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/faaad20fe6ce31f428d496619c633f35/logTestPromoteNonVoterInAddVoter4102867872
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:146: using test tenant configuration from testserver factory defaults
    test_server_shim.go:637: DRPC is enabled (via testserver factory defaults)
    replicate_queue_test.go:2286: success: 84 has 3 voters and 2 non-voters
    replicate_queue_test.go:2362: 
        	Error Trace:	pkg/kv/kvserver/replicate_queue_test.go:2362
        	Error:      	Expected to promote non-voter, instead added voter
        	Test:       	TestPromoteNonVoterInAddVoter
        	Messages:   	Added voter store ID: 6
        	            	Add voter events: [{r84:/{Table/104-Max} [(n1,s1):1, (n2,s2):5, (n3,s3):7, (n5,s5):8, (n7,s7):9NON_VOTER, next=10, gen=28] <nil> <nil> (n5,s5):8 <nil> range under-replicated {"Target":"s5, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:65"}} {r84:/{Table/104-Max} [(n1,s1):1, (n2,s2):5, (n3,s3):7, (n5,s5):8, (n7,s7):9NON_VOTER, (n6,s6):10LEARNER, next=11, gen=29] <nil> <nil> (n6,s6):10LEARNER <nil> range under-replicated {"Target":"s6, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:63"}}]
    exectrace.go:39: execution trace written to /artifacts/tmp/_tmp/faaad20fe6ce31f428d496619c633f35/logTestPromoteNonVoterInAddVoter4102867872/exectrace_goid_59698.bin
    panic.go:694: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/faaad20fe6ce31f428d496619c633f35/logTestPromoteNonVoterInAddVoter4102867872
--- FAIL: TestPromoteNonVoterInAddVoter (368.14s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPromoteNonVoterInAddVoter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63889

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -2126,9 +2126,19 @@ func TestPromoteNonVoterInAddVoter(t *testing.T) {
 	// SQL update, changing the span configs (which should register on all nodes),
 	// and subsequent replication changes. In failing runs, it will be interesting
 	// which prefix of events remains.
+	//
+	// `allocator_scorer=3` logs per-candidate scoring details (diversity,
+	// balance, convergence, range counts) when the allocator picks a target. This
+	// is useful for diagnosing the likely failure mode below, see #170221: when
+	// the balance scores of the existing non-voter and a fresh voter candidate
+	// diverge (e.g. because their range counts differ enough to land in
+	// different balance buckets), the allocator's promotion preference for the
+	// existing non-voter is not applied, and a brand new voter is added instead.
+	// The per-candidate scoring output makes it possible to see which inputs
+	// pushed the two candidates into different balance buckets.
 	{
 		old := log.GetVModule()
-		changed := "store=2,reconciler=3"
+		changed := "store=2,reconciler=3,allocator_scorer=3"
 		if old != "" {
 			changed = old + "," + changed
 		}

